File: cqc/cLib/testSend.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>

// #include <stdlib.h>

#include <netinet/in.h>
#include <netdb.h>

#include <string.h>
#include<arpa/inet.h>
 

#include "cqc.h"

int main(int argc, char *argv[]) {

	uint16_t portno;
	char *hostname;
	uint16_t remotePort;
	char *remoteHost;
	struct in_addr remoteNode;
	cqc_lib *cqc;
	int app_id;
	int outcome;
	struct hostent *server;
	uint16_t qubit;

	/* Retrieve arguments from command line */
   	if (argc < 5) {
      		fprintf(stderr,"usage %s hostname port remoteHost remotePort\n", argv[0]);
      		exit(0);
   	}
	hostname = argv[1];	
   	portno = atoi(argv[2]);
	remoteHost = argv[3];
	remotePort = atoi(argv[4]);

	/* Lookup remote host */
        server = gethostbyname(remoteHost);
        if (server == NULL) {
                fprintf(stderr,"ERROR, no such host\n");
                return(-1);
        }

	/* In this example, we are simply application 10 */
	app_id = 10;

	/* In this example, we will not check for errors. All functions return -1 on failure */
	cqc = cqc_init(app_id);
	cqc_connect(cqc, hostname, portno);

	cqc_simple_cmd(cqc, CQC_CMD_NEW, 0, 0);
	qubit = cqc_wait_until_newok(cqc);

	remoteNode.s_addr = ntohl(*((uint32_t *)server->h_addr));
	cqc_send(cqc, qubit, app_id, remoteNode.s_addr, remotePort);
	cqc_wait_until_done(cqc, 1);

   	return 0;
}

File: cqc/rustLib/tests/test_send.rs
extern crate rust_lib;

use rust_lib::Cqc;
use rust_lib::cqc_api::*;
use std::net::Ipv4Addr;

#[test]
fn test_send() {
    // Retrieve arguments from command line
    // These four variables should be read from somewhere else
    let hostname = String::from("localhost");
    let portno: u16 = 8821;
    let remote_host: u32 = u32::from(Ipv4Addr::new(127, 0, 0, 1));
    let remote_port: u16 = 8822;

    // In this example, we are simply application 10
    let app_id: u16 = 10;

    // In this example, we will not check for errors.
    // Initialise a CQC service.
    let cqc = Cqc::new(app_id, &hostname, portno).unwrap();

    cqc.simple_cmd(CQC_CMD_NEW, 0, false).unwrap();
    let qubit = cqc.wait_until_newok().unwrap();

    cqc.send(qubit, app_id, remote_host, remote_port).unwrap();
    cqc.wait_until_done(1).unwrap();
}
